core/jobs/types/model_property.py

```diff
@@ -109,62 +134,99 @@ def yield_value_from_model(self, model):
         else:
             yield value
 
-    def _to_model_class(self):
+    def _to_model_class(self) -> Type[base_models.BaseModel]:
         """Returns the model class associated with this instance.
 
         Returns:
             type(BaseModel). The model type.
         """
-        return job_utils.get_model_class(self._model_kind)
+        model_class = job_utils.get_model_class(self._model_kind)
+
+        # To narrow down the type from datastore_services.Model to
+        # base_models.BaseModel, we used assert statement here.
+        assert issubclass(model_class, base_models.BaseModel)
+        return model_class
 
-    def _to_property(self):
+    def _to_property(self) -> PropertyType:
         """Returns the Property object associated with this instance.
 
         Returns:
             *. A property instance.
         """
-        return getattr(self._to_model_class(), self._property_name)
+        property_obj = getattr(self._to_model_class(), self._property_name)
+
+        # The behavior of `id` Python property is different during type
+        # checking and during runtime. During type checking it is considered as
+        # `Callable[]` because a Python property is decorated using Python's
+        # property class, while during runtime a Python property is considered
+        # as instance of property class. So to split the assertion in both the
+        # cases, we used `if MYPY:` clause here.
+        if MYPY: # pragma: no cover
+            assert (
+                isinstance(property_obj, datastore_services.Property) and
+                callable(property_obj)
+            )
+        else:
+            assert isinstance(
+                property_obj,
+                (datastore_services.Property, property)
+            )
```

Comment:
```suggestion
        assert isinstance(property_obj, datastore_services.Property)
        if MYPY: # pragma: no cover
            assert callable(property_obj)
        else:
            assert isinstance(property_obj, property)
```


What is `property` here? I cannot find it anywhere.

Reply:
Here, `property`  is python's inbuilt property class.

Reply:
Also, this suggestion fails backend tests because:

Consider a case where `property_obj` is `id` python property, then in that case `assert isinstance(property_obj, datastore_services.Property)` stops the execution and throws an assertion error. because this suggestion always checks for `datastore_services.Property` first then it passes to other checks.

Reply:
Okay, makes sense.